### PR TITLE
#3138 improved coverage for sharding core preprocessor

### LINKED ISSUE #3138: Add more unit test cases for sharding-core-preprocessor module
sharding-core-preprocessor module is decouple with Sharding & encrypt. The codes has already stable.
Should Add more unit test cases for this module.

### PR DESCRIPTION
**Pull request by sluk3r**
Fixes #3138

the following test cases added:
 
-ShorthandProjectionTest
-TablesContext#getSchema, TablesContext#assertInstanceCreatedWhenNoExceptionThrown
-ProjectionsContextTest#assertGetColumnLabelsWhenGetQualifiedShorthandColumnLabelsResultEmpty

### CODE REVIEW

File: sharding-core/sharding-core-preprocessor/src/test/java/org/apache/shardingsphere/core/preprocessor/segment/table/TablesContextTest.java [hunk 1]
```diff
@@ -181,6 +185,24 @@ public void assertFindTableNameWhenColumnSegmentOwnerAbsentAndTableMetasContains
...
+    }
+    
+    @Test
+    public void assertInstanceCreatedWhenNoExceptionThrown() {
+        SQLStatement sqlStatement = mock(SQLStatement.class);
+        TableSegment tableSegment = new TableSegment(0, 10, "TableSegmentName");
+        SchemaSegment schemaSegment = mock(SchemaSegment.class);
+        when(schemaSegment.getName()).thenReturn("SchemaSegmentName");
+        tableSegment.setOwner(schemaSegment);
+        when(sqlStatement.findSQLSegments(TableAvailable.class)).thenReturn(Collections.singletonList((TableAvailable) tableSegment));
```
Comment (terrymanu) [inline sharding-core/sharding-core-preprocessor/src/test/java/org/apache/shardingsphere/core/preprocessor/segment/table/TablesContextTest.java]: More blank line is unnecessary.

File: sharding-core/sharding-core-preprocessor/src/test/java/org/apache/shardingsphere/core/preprocessor/segment/select/projection/engine/ProjectionsContextEngineTest.java [hunk 2]
```diff
@@ -59,4 +62,40 @@ public void assertProjectionsContextCreatedProperlyWhenSelectItemPresent() {
...
+        SelectItemsSegment selectItemsSegment = mock(SelectItemsSegment.class);
+        when(selectStatement.getSelectItems()).thenReturn(selectItemsSegment);
+        ShorthandSelectItemSegment shorthandSelectItemSegment = new ShorthandSelectItemSegment(0, 10, "text");
+        TableSegment owner = new TableSegment(0, 10, "name");
+        owner.setOwner(new SchemaSegment(0, 10, "name"));
+        shorthandSelectItemSegment.setOwner(owner);
+        when(selectItemsSegment.getSelectItems()).thenReturn(Collections.<SelectItemSegment>singleton(shorthandSelectItemSegment));
+        OrderByContext orderByContext = mock(OrderByContext.class);
+        OrderByItem orderByItem = mock(OrderByItem.class);
+        when(orderByItem.getSegment()).thenReturn(mock(IndexOrderByItemSegment.class));
```
Comment (terrymanu) [inline sharding-core/sharding-core-preprocessor/src/test/java/org/apache/shardingsphere/core/preprocessor/segment/select/projection/engine/ProjectionsContextEngineTest.java]: More blank line is unnecessary.

Comment (terrymanu) [inline sharding-core/sharding-core-preprocessor/src/test/java/org/apache/shardingsphere/core/preprocessor/segment/select/projection/engine/ProjectionsContextEngineTest.java]: More blank line is unnecessary.

File: sharding-core/sharding-core-preprocessor/src/test/java/org/apache/shardingsphere/core/preprocessor/segment/select/projection/engine/ProjectionsContextEngineTest.java [hunk 3]
```diff
@@ -59,4 +62,40 @@ public void assertProjectionsContextCreatedProperlyWhenSelectItemPresent() {
         ProjectionsContext projectionsContext = new ProjectionsContextEngine(null).createProjectionsContext(null, selectStatement, mock(GroupByContext.class), mock(OrderByContext.class));
         assertNotNull(projectionsContext);
     }
+    
+    @Test
+    public void assertProjectionsContextCreatedProperlyWhenOrderByContextOrderItemsPresent() {
```
Comment (terrymanu) [inline sharding-core/sharding-core-preprocessor/src/test/java/org/apache/shardingsphere/core/preprocessor/segment/select/projection/engine/ProjectionsContextEngineTest.java]: The test method name should same with to be tested method name, so the better name should be `assert createProjectionsContextWhenOrderByContextOrderItemsPresent`

File: sharding-core/sharding-core-preprocessor/src/test/java/org/apache/shardingsphere/core/preprocessor/segment/select/projection/engine/ProjectionsContextEngineTest.java [hunk 4]
```diff
@@ -59,4 +62,40 @@ public void assertProjectionsContextCreatedProperlyWhenSelectItemPresent() {
...
+        ShorthandSelectItemSegment shorthandSelectItemSegment = new ShorthandSelectItemSegment(0, 10, "text");
+        TableSegment owner = new TableSegment(0, 10, "name");
+        owner.setOwner(new SchemaSegment(0, 10, "name"));
+        shorthandSelectItemSegment.setOwner(owner);
+        when(selectItemsSegment.getSelectItems()).thenReturn(Collections.<SelectItemSegment>singleton(shorthandSelectItemSegment));
+        OrderByContext orderByContext = mock(OrderByContext.class);
+        OrderByItem orderByItem = mock(OrderByItem.class);
+        when(orderByItem.getSegment()).thenReturn(mock(IndexOrderByItemSegment.class));
+        when(orderByContext.getItems()).thenReturn(Collections.singletonList(orderByItem));
+        ProjectionsContext projectionsContext = new ProjectionsContextEngine(null).createProjectionsContext(null, selectStatement, mock(GroupByContext.class), orderByContext);
```
Comment (terrymanu) [inline sharding-core/sharding-core-preprocessor/src/test/java/org/apache/shardingsphere/core/preprocessor/segment/select/projection/engine/ProjectionsContextEngineTest.java]: The name of to be tested variable should be `actual`

File: sharding-core/sharding-core-preprocessor/src/test/java/org/apache/shardingsphere/core/preprocessor/segment/select/projection/impl/ShorthandProjectionTest.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+import org.junit.Test;
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+public final class ShorthandProjectionTest {
+    
+    @Test
+    public void assertGetExpression() {
+        assertTrue(new ShorthandProjection("owner").getExpression().contains(".*"));
```
Comment (terrymanu) [inline sharding-core/sharding-core-preprocessor/src/test/java/org/apache/shardingsphere/core/preprocessor/segment/select/projection/impl/ShorthandProjectionTest.java]: We need the assertion more accurately, just assert full text of expression

File: sharding-core/sharding-core-preprocessor/src/test/java/org/apache/shardingsphere/core/preprocessor/segment/select/projection/impl/ShorthandProjectionTest.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    }
+    
+    @Test
+    public void assertGetAlias() {
```
Comment (terrymanu) [inline sharding-core/sharding-core-preprocessor/src/test/java/org/apache/shardingsphere/core/preprocessor/segment/select/projection/impl/ShorthandProjectionTest.java]: The method name should be `assertGetAliasWhenAbsent`
And here are missing assert for `assertGetAliasWhenPresent`

Comment (sluk3r) [reply sharding-core/sharding-core-preprocessor/src/test/java/org/apache/shardingsphere/core/preprocessor/segment/select/projection/impl/ShorthandProjectionTest.java]: hi, 
I think the name of assertGetAliasWhenAbsent is ok. 

but as for "here are missing assert for assertGetAliasWhenPresent", It seems that it is not necessary, as the whole method block under test is: 

```java
   @Override
    public Optional<String> getAlias() {
        return Optional.absent();
    }
```

so It seems that there is no possible the method would return Optional.present().

File: sharding-core/sharding-core-preprocessor/src/test/java/org/apache/shardingsphere/core/preprocessor/segment/table/TablesContextTest.java (hunk 1 above)
Comment (sluk3r) [reply sharding-core/sharding-core-preprocessor/src/test/java/org/apache/shardingsphere/core/preprocessor/segment/table/TablesContextTest.java]: It seems that there is no blank lines?

File: sharding-core/sharding-core-preprocessor/src/test/java/org/apache/shardingsphere/core/preprocessor/segment/select/projection/engine/ProjectionsContextEngineTest.java (hunk 2 above)
Comment (sluk3r) [reply sharding-core/sharding-core-preprocessor/src/test/java/org/apache/shardingsphere/core/preprocessor/segment/select/projection/engine/ProjectionsContextEngineTest.java]: same as the above

File: sharding-core/sharding-core-preprocessor/src/test/java/org/apache/shardingsphere/core/preprocessor/segment/select/projection/impl/ShorthandProjectionTest.java (hunk 6 above)
Comment (terrymanu) [reply sharding-core/sharding-core-preprocessor/src/test/java/org/apache/shardingsphere/core/preprocessor/segment/select/projection/impl/ShorthandProjectionTest.java]: OK

File: sharding-core/sharding-core-preprocessor/src/test/java/org/apache/shardingsphere/core/preprocessor/segment/select/projection/engine/ProjectionsContextEngineTest.java [hunk 7]
```diff
@@ -59,4 +62,40 @@ public void assertProjectionsContextCreatedProperlyWhenSelectItemPresent() {
...
+        TableSegment owner = new TableSegment(0, 10, "name");
+        owner.setOwner(new SchemaSegment(0, 10, "name"));
+        shorthandSelectItemSegment.setOwner(owner);
+        when(selectItemsSegment.getSelectItems()).thenReturn(Collections.<SelectItemSegment>singleton(shorthandSelectItemSegment));
+        OrderByContext orderByContext = mock(OrderByContext.class);
+        OrderByItem orderByItem = mock(OrderByItem.class);
+        when(orderByItem.getSegment()).thenReturn(mock(TextOrderByItemSegment.class));
+        when(orderByContext.getItems()).thenReturn(Collections.singletonList(orderByItem));
+        ProjectionsContext projectionsContext = new ProjectionsContextEngine(null).createProjectionsContext(null, selectStatement, mock(GroupByContext.class), orderByContext);
+        assertNotNull(projectionsContext);
```
Comment (terrymanu) [inline sharding-core/sharding-core-preprocessor/src/test/java/org/apache/shardingsphere/core/preprocessor/segment/select/projection/engine/ProjectionsContextEngineTest.java]: The name of to be tested variable should be `actual`

File: sharding-core/sharding-core-preprocessor/src/test/java/org/apache/shardingsphere/core/preprocessor/segment/select/projection/engine/ProjectionsContextEngineTest.java [hunk 8]
```diff
@@ -59,4 +62,40 @@ public void assertProjectionsContextCreatedProperlyWhenSelectItemPresent() {
...
+        OrderByContext orderByContext = mock(OrderByContext.class);
+        OrderByItem orderByItem = mock(OrderByItem.class);
+        when(orderByItem.getSegment()).thenReturn(mock(IndexOrderByItemSegment.class));
+        when(orderByContext.getItems()).thenReturn(Collections.singletonList(orderByItem));
+        ProjectionsContext actual = new ProjectionsContextEngine(null).createProjectionsContext(null, selectStatement, mock(GroupByContext.class), orderByContext);
+        assertNotNull(actual);
+    }
+    
+    @Test
+    public void assertProjectionsContextCreatedProperlyWhenOrderByItemSegmentNotInstanceOfIndexOrderByItemSegment() {
```
Comment (terrymanu) [inline sharding-core/sharding-core-preprocessor/src/test/java/org/apache/shardingsphere/core/preprocessor/segment/select/projection/engine/ProjectionsContextEngineTest.java]: Why the method name as `assertProjectionsContextCreated...`?
The method to be tested name is `createProjectionsContext`, the test method should keep consist with be tested one.

Comment (sluk3r) [reply sharding-core/sharding-core-preprocessor/src/test/java/org/apache/shardingsphere/core/preprocessor/segment/select/projection/engine/ProjectionsContextEngineTest.java]: the method name of assertProjectionsContextCreatedProperlyWhenOrderByItemSegmentNotInstanceOfIndexOrderByItemSegment express the meaning: assert that ProjectionsContext instance is created when OrderByItemSegment is not instance of InstanceOfIndexOrderByItemSegment, corresponding to the following switch in the format of negation:

image


So, I think the method name of assertProjectionsContextCreatedProperlyWhenOrderByItemSegmentNotInstanceOfIndexOrderByItemSegment could not be changed.

Comment (terrymanu) [reply sharding-core/sharding-core-preprocessor/src/test/java/org/apache/shardingsphere/core/preprocessor/segment/select/projection/engine/ProjectionsContextEngineTest.java]: I don't think so. In the unit test conduct, the testing method's prefix should name as the to be tested method which to let user know what is actual testing method of this case.
The proposal of the test case is that test the method `createProjectionsContext`, so the test case name should be start with `assertCreateProjectionsContext`.

`assertCreateProjectionsContextWithoutIndexOrderByItemSegment` maybe a better name.

Comment (sluk3r) [reply sharding-core/sharding-core-preprocessor/src/test/java/org/apache/shardingsphere/core/preprocessor/segment/select/projection/engine/ProjectionsContextEngineTest.java]: OK

**Pull request merged**